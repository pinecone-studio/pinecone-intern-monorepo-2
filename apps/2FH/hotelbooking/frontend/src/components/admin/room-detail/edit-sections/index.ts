export { BasicInfoSection } from './BasicInfoSection';
export { AmenitiesSection } from './AmenitiesSection';
export { ImagesSection } from './ImagesSection';
export { DetailsSection } from './DetailsSection';


