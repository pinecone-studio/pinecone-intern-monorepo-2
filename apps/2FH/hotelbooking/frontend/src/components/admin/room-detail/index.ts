export { RoomInfoCard } from './RoomInfoCard';
export { RoomAmenitiesCard } from './RoomAmenitiesCard';
export { RoomImagesCard } from './RoomImagesCard';
export { LoadingSkeleton } from './LoadingSkeleton';
export { ErrorMessage } from './ErrorMessage';
export { NotFoundMessage } from './NotFoundMessage';
export { RoomDetailPage } from './RoomDetailPage';
export { EditRoomModal } from './EditRoomModal';
