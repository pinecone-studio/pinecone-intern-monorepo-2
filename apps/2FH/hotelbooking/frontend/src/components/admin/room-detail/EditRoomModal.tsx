/* eslint-disable  */
'use client';
import { useState } from 'react';
import { useUpdateRoomMutation, Status, Response, RoomUpdateInput } from '@/generated';
import { Button } from '@/components/ui/button';
import { Dialog, DialogContent, DialogHeader, DialogTitle, DialogFooter } from '@/components/ui/dialog';
import { BasicInfoSection, AmenitiesSection, ImagesSection, DetailsSection } from './edit-sections';

interface EditRoomModalProps {
  room: any;
  section: 'basic' | 'amenities' | 'images' | 'details';
  isOpen: boolean;
  onOpenChange: (open: boolean) => void;
  refetch: () => Promise<any>;
  roomId: string;
}

export const EditRoomModal = ({ room, section, isOpen, onOpenChange, refetch, roomId }: EditRoomModalProps) => {
  const [updateRoom, { loading: updateLoading }] = useUpdateRoomMutation();
  const [error, setError] = useState<string | null>(null);

  const normalizeStatus = (status: any): Status => {
    if (!status) return Status.Available;
    const statusStr = status.toString();

    // Map common variations to correct enum values
    switch (statusStr.toLowerCase()) {
      case 'available':
        return Status.Available;
      case 'booked':
        return Status.Booked;
      case 'cancelled':
        return Status.Cancelled;
      case 'completed':
        return Status.Completed;
      case 'pending':
        return Status.Pending;
      default:
        // If it's already a valid enum value, return it
        if (Object.values(Status).includes(statusStr as Status)) {
          return statusStr as Status;
        }
        return Status.Available;
    }
  };

  const [formData, setFormData] = useState(() => ({
    ...room,
    status: normalizeStatus(room.status),
    internet: room.internet || [],
    foodAndDrink: room.foodAndDrink || [],
    bedRoom: room.bedRoom || [],
    bathroom: room.bathroom || [],
    accessibility: room.accessibility || [],
    entertainment: room.entertainment || [],
    other: room.other || [],
    roomInformation: room.roomInformation || [],
    imageURL: room.imageURL || [],
  }));

  const handleInputChange = (field: string, value: any) => {
    setFormData((prev: any) => ({
      ...prev,
      [field]: value,
    }));
  };

  const handleSave = async () => {
    try {
      setError(null);

      const input: RoomUpdateInput = {
        bedNumber: formData.bedNumber || room.bedNumber,
        status: normalizeStatus(formData.status || room.status),
      };

      if (section === 'basic') {
        if (formData.name !== undefined) input.name = formData.name;
        if (formData.pricePerNight !== undefined) input.pricePerNight = formData.pricePerNight;
        if (formData.typePerson !== undefined) input.typePerson = formData.typePerson;
        if (formData.roomInformation !== undefined) input.roomInformation = formData.roomInformation;
      } else if (section === 'amenities') {
        if (formData.internet !== undefined) input.internet = formData.internet;
        if (formData.foodAndDrink !== undefined) input.foodAndDrink = formData.foodAndDrink;
        if (formData.bedRoom !== undefined) input.bedRoom = formData.bedRoom;
        if (formData.bathroom !== undefined) input.bathroom = formData.bathroom;
        if (formData.accessibility !== undefined) input.accessibility = formData.accessibility;
        if (formData.entertainment !== undefined) input.entertainment = formData.entertainment;
        if (formData.other !== undefined) input.other = formData.other;
      } else if (section === 'images') {
        if (formData.imageURL !== undefined) input.imageURL = formData.imageURL;
      } else if (section === 'details') {
        if (formData.roomInformation !== undefined) input.roomInformation = formData.roomInformation;
      }

      const result = await updateRoom({
        variables: {
          updateRoomId: roomId,
          input,
        },
      });

      if (result.data?.updateRoom === Response.Success) {
        await refetch();
        onOpenChange(false);
      } else {
        setError('Failed to update room. Please try again.');
      }
    } catch (error: any) {
      console.error('Error updating room:', error);
      setError('An error occurred while updating the room. Please try again.');
    }
  };
  const getSectionComponent = () => {
    let component;
    switch (section) {
      case 'basic':
        component = <BasicInfoSection room={formData} handleInputChange={handleInputChange} />;
        break;
      case 'amenities':
        component = <AmenitiesSection room={formData} handleInputChange={handleInputChange} />;
        break;
      case 'images':
        component = <ImagesSection room={formData} handleInputChange={handleInputChange} />;
        break;
      case 'details':
        component = <DetailsSection room={formData} handleInputChange={handleInputChange} />;
        break;
      default:
        component = null;
    }
    return component || null;
  };
  const getSectionTitle = () => {
    const titleMap = {
      basic: 'Basic Information',
      amenities: 'Amenities',
      images: 'Images',
      details: 'Details',
    };
    return titleMap[section] || 'Edit Room';
  };
  const handleOpenChange = (newOpen: boolean) => {
    onOpenChange(newOpen);
    if (newOpen) {
      setError(null);
    }
  };

  return (
    <Dialog open={isOpen} onOpenChange={handleOpenChange}>
      <DialogContent className="max-w-4xl max-h-[90vh] overflow-y-auto" data-testid="edit-room-modal">
        <DialogHeader>
          <DialogTitle data-testid="edit-room-modal-title">Edit {getSectionTitle()}</DialogTitle>
        </DialogHeader>

        <div className="space-y-6 py-4">
          {error && (
            <div className="bg-red-50 border border-red-200 rounded-md p-3">
              <p className="text-sm text-red-600">{error}</p>
            </div>
          )}
          <div className="min-h-[400px]" data-testid="edit-room-modal-content">
            {getSectionComponent()}
          </div>
        </div>

        <DialogFooter className="flex justify-end gap-3 pt-4 border-t">
          <Button variant="outline" onClick={() => onOpenChange(false)} disabled={updateLoading} data-testid="edit-room-modal-cancel">
            Cancel
          </Button>
          <Button onClick={handleSave} className="bg-blue-600 hover:bg-blue-700" disabled={updateLoading} data-testid="edit-room-modal-save">
            {updateLoading ? 'Saving...' : 'Save Changes'}
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
};
