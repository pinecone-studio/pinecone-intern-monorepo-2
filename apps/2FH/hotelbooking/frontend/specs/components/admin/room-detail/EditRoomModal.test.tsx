/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { EditRoomModal } from '@/components/admin/room-detail/EditRoomModal';
import { MockedProvider } from '@apollo/client/testing';

// Mock the generated GraphQL hooks
const mockUpdateRoom = jest.fn();
jest.mock('@/generated', () => ({
  useUpdateRoomMutation: () => [mockUpdateRoom, { loading: false, error: null, data: { updateRoom: { id: 'room-1' } } }],
}));

// Mock UI components
jest.mock('@/components/ui/button', () => ({
  Button: ({ children, onClick, variant, className, disabled, ...props }: any) => (
    <button onClick={onClick} variant={variant} className={className} disabled={disabled} {...props}>
      {children}
    </button>
  ),
}));

jest.mock('@/components/ui/dialog', () => ({
  Dialog: ({ children, open, onOpenChange }: any) => (
    <div data-testid="dialog" data-open={open}>
      {children}
    </div>
  ),
  DialogContent: ({ children, className, ...props }: any) => (
    <div className={className} {...props}>
      {children}
    </div>
  ),
  DialogHeader: ({ children }: any) => <div data-testid="dialog-header">{children}</div>,
  DialogTitle: ({ children, ...props }: any) => (
    <div data-testid="dialog-title" {...props}>
      {children}
    </div>
  ),
  DialogFooter: ({ children, className }: any) => (
    <div data-testid="dialog-footer" className={className}>
      {children}
    </div>
  ),
}));

// Mock the edit sections
jest.mock('@/components/admin/room-detail/edit-sections', () => ({
  BasicInfoSection: ({ room, handleInputChange }: any) => (
    <div data-testid="basic-info-section">
      <input data-testid="name-input" value={room.name || ''} onChange={(e) => handleInputChange('name', e.target.value)} placeholder="Room name" />
      <input data-testid="price-input" value={room.pricePerNight || ''} onChange={(e) => handleInputChange('pricePerNight', e.target.value)} placeholder="Price per night" />
      <input data-testid="type-person-input" value={room.typePerson || ''} onChange={(e) => handleInputChange('typePerson', e.target.value)} placeholder="Type person" />
    </div>
  ),
  AmenitiesSection: ({ room, handleInputChange }: any) => (
    <div data-testid="amenities-section">
      <input data-testid="internet-input" value={room.internet || ''} onChange={(e) => handleInputChange('internet', e.target.value)} placeholder="Internet" />
    </div>
  ),
  ImagesSection: ({ room, handleInputChange }: any) => (
    <div data-testid="images-section">
      <input data-testid="image-url-input" value={room.imageURL || ''} onChange={(e) => handleInputChange('imageURL', e.target.value)} placeholder="Image URL" />
    </div>
  ),
  DetailsSection: ({ room, handleInputChange }: any) => (
    <div data-testid="details-section">
      <textarea data-testid="room-information-input" value={room.roomInformation || ''} onChange={(e) => handleInputChange('roomInformation', e.target.value)} placeholder="Room information" />
    </div>
  ),
}));

describe('EditRoomModal', () => {
  const mockRoom = {
    id: 'room-1',
    name: 'Deluxe Suite',
    pricePerNight: 150000,
    typePerson: 4,
    bedNumber: 2,
    status: 'available',
    internet: ['Free WiFi'],
    foodAndDrink: ['Mini Bar'],
    bedRoom: ['King Bed'],
    bathroom: ['Shower'],
    accessibility: ['Wheelchair Accessible'],
    entertainment: ['TV'],
    other: ['Room Service'],
    roomInformation: ['Free WiFi', 'Air Conditioning'],
    imageURL: ['https://example.com/image1.jpg'],
  };

  const defaultProps = {
    room: mockRoom,
    section: 'basic' as const,
    isOpen: true,
    onOpenChange: jest.fn(),
    refetch: jest.fn(),
    roomId: 'room-1',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    mockUpdateRoom.mockResolvedValue({
      data: { updateRoom: { id: 'room-1' } },
    });
  });

  const renderWithMocks = (props = defaultProps) => {
    return render(
      <MockedProvider mocks={[]} addTypename={false}>
        <EditRoomModal {...props} />
      </MockedProvider>
    );
  };

  describe('Rendering', () => {
    it('renders modal when open', () => {
      renderWithMocks();

      expect(screen.getByTestId('dialog')).toHaveAttribute('data-open', 'true');
      expect(screen.getByTestId('edit-room-modal')).toBeInTheDocument();
      expect(screen.getByTestId('dialog-header')).toBeInTheDocument();
      expect(screen.getByTestId('dialog-footer')).toBeInTheDocument();
    });

    it('does not render modal when closed', () => {
      renderWithMocks({ ...defaultProps, isOpen: false });

      expect(screen.getByTestId('dialog')).toHaveAttribute('data-open', 'false');
    });

    it('renders correct title for basic section', () => {
      renderWithMocks();

      expect(screen.getByTestId('edit-room-modal-title')).toHaveTextContent('Edit Basic Information');
    });

    it('renders correct title for amenities section', () => {
      renderWithMocks({ ...defaultProps, section: 'amenities' });

      expect(screen.getByTestId('edit-room-modal-title')).toHaveTextContent('Edit Amenities');
    });

    it('renders correct title for images section', () => {
      renderWithMocks({ ...defaultProps, section: 'images' });

      expect(screen.getByTestId('edit-room-modal-title')).toHaveTextContent('Edit Images');
    });

    it('renders correct title for details section', () => {
      renderWithMocks({ ...defaultProps, section: 'details' });

      expect(screen.getByTestId('edit-room-modal-title')).toHaveTextContent('Edit Details');
    });

    it('renders default title for unknown section', () => {
      renderWithMocks({ ...defaultProps, section: 'unknown' as any });

      expect(screen.getByTestId('edit-room-modal-title')).toHaveTextContent('Edit Edit Room');
    });
  });

  describe('Section Rendering', () => {
    it('renders BasicInfoSection for basic section', () => {
      renderWithMocks();

      expect(screen.getByTestId('basic-info-section')).toBeInTheDocument();
      expect(screen.getByTestId('name-input')).toBeInTheDocument();
      expect(screen.getByTestId('price-input')).toBeInTheDocument();
      expect(screen.getByTestId('type-person-input')).toBeInTheDocument();
    });

    it('renders AmenitiesSection for amenities section', () => {
      renderWithMocks({ ...defaultProps, section: 'amenities' });

      expect(screen.getByTestId('amenities-section')).toBeInTheDocument();
      expect(screen.getByTestId('internet-input')).toBeInTheDocument();
    });

    it('renders ImagesSection for images section', () => {
      renderWithMocks({ ...defaultProps, section: 'images' });

      expect(screen.getByTestId('images-section')).toBeInTheDocument();
      expect(screen.getByTestId('image-url-input')).toBeInTheDocument();
    });

    it('renders DetailsSection for details section', () => {
      renderWithMocks({ ...defaultProps, section: 'details' });

      expect(screen.getByTestId('details-section')).toBeInTheDocument();
      expect(screen.getByTestId('room-information-input')).toBeInTheDocument();
    });

    it('renders null for unknown section', () => {
      renderWithMocks({ ...defaultProps, section: 'unknown' as any });

      expect(screen.queryByTestId('basic-info-section')).not.toBeInTheDocument();
      expect(screen.queryByTestId('amenities-section')).not.toBeInTheDocument();
      expect(screen.queryByTestId('images-section')).not.toBeInTheDocument();
      expect(screen.queryByTestId('details-section')).not.toBeInTheDocument();
    });
  });

  describe('Form State Management', () => {
    it('initializes form data with room data', () => {
      renderWithMocks();

      expect(screen.getByTestId('name-input')).toHaveValue('Deluxe Suite');
      expect(screen.getByTestId('price-input')).toHaveValue('150000');
      expect(screen.getByTestId('type-person-input')).toHaveValue('4');
    });

    it('handles room with null/undefined values', () => {
      const roomWithNulls = {
        id: 'room-1',
        name: null,
        pricePerNight: undefined,
        typePerson: null,
        status: null,
        internet: null,
        foodAndDrink: undefined,
        bedRoom: null,
        bathroom: undefined,
        accessibility: null,
        entertainment: undefined,
        other: null,
        roomInformation: null,
        imageURL: undefined,
      };

      renderWithMocks({ ...defaultProps, room: roomWithNulls });

      expect(screen.getByTestId('name-input')).toHaveValue('');
      expect(screen.getByTestId('price-input')).toHaveValue('');
      expect(screen.getByTestId('type-person-input')).toHaveValue('');
    });

    it('handles input changes correctly', async () => {
      const user = userEvent.setup();
      renderWithMocks();

      const nameInput = screen.getByTestId('name-input');
      await user.clear(nameInput);
      await user.type(nameInput, 'Updated Room Name');

      expect(nameInput).toHaveValue('Updated Room Name');
    });

    it('normalizes status correctly', () => {
      const roomWithInvalidStatus = {
        ...mockRoom,
        status: 'INVALID_STATUS',
      };

      renderWithMocks({ ...defaultProps, room: roomWithInvalidStatus });

      // The component should normalize the status to 'available'
      expect(screen.getByTestId('dialog')).toBeInTheDocument();
    });

    it('handles status normalization for different values', () => {
      const statuses = ['BOOKED', 'booked', 'CANCELLED', 'cancelled', 'PENDING', 'pending', 'COMPLETED', 'completed', 'AVAILABLE', 'available'];

      statuses.forEach((status) => {
        const roomWithStatus = { ...mockRoom, status };
        const { rerender } = renderWithMocks({ ...defaultProps, room: roomWithStatus });
        expect(screen.getByTestId('dialog')).toBeInTheDocument();
        rerender(<div />);
      });
    });
  });

  describe('Save Functionality', () => {
    it('calls updateRoom mutation with correct variables for basic section', async () => {
      const user = userEvent.setup();
      renderWithMocks();

      const nameInput = screen.getByTestId('name-input');
      await user.clear(nameInput);
      await user.type(nameInput, 'Updated Name');

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              bedNumber: 2,
              status: 'available',
              name: 'Updated Name',
              pricePerNight: 150000,
              roomInformation: ['Free WiFi', 'Air Conditioning'],
              typePerson: 4,
            },
          },
        });
      });
    });

    it('calls updateRoom mutation with correct variables for amenities section', async () => {
      const user = userEvent.setup();
      renderWithMocks({ ...defaultProps, section: 'amenities' });

      const internetInput = screen.getByTestId('internet-input');
      await user.clear(internetInput);
      await user.type(internetInput, 'Updated Internet');

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              accessibility: ['Wheelchair Accessible'],
              bathroom: ['Shower'],
              bedNumber: 2,
              bedRoom: ['King Bed'],
              entertainment: ['TV'],
              foodAndDrink: ['Mini Bar'],
              internet: 'Updated Internet',
              other: ['Room Service'],
              status: 'available',
            },
          },
        });
      });
    });

    it('calls updateRoom mutation with correct variables for images section', async () => {
      const user = userEvent.setup();
      renderWithMocks({ ...defaultProps, section: 'images' });

      const imageInput = screen.getByTestId('image-url-input');
      await user.clear(imageInput);
      await user.type(imageInput, 'Updated Image URL');

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              bedNumber: 2,
              imageURL: 'Updated Image URL',
              status: 'available',
            },
          },
        });
      });
    });

    it('calls updateRoom mutation with correct variables for details section', async () => {
      const user = userEvent.setup();
      renderWithMocks({ ...defaultProps, section: 'details' });

      const roomInfoInput = screen.getByTestId('room-information-input');
      await user.clear(roomInfoInput);
      await user.type(roomInfoInput, 'Updated Room Information');

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              bedNumber: 2,
              roomInformation: 'Updated Room Information',
              status: 'available',
            },
          },
        });
      });
    });

    it('calls refetch and closes modal on successful save', async () => {
      const user = userEvent.setup();
      renderWithMocks();

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(defaultProps.refetch).toHaveBeenCalled();
        expect(defaultProps.onOpenChange).toHaveBeenCalledWith(false);
      });
    });

    it('handles save error gracefully', async () => {
      const user = userEvent.setup();
      mockUpdateRoom.mockRejectedValue(new Error('Save failed'));

      // Mock console.error to avoid noise in test output
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

      renderWithMocks();

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith('Error updating room:', expect.any(Error));
      });

      consoleSpy.mockRestore();
    });

    it('handles save with no data returned', async () => {
      const user = userEvent.setup();
      mockUpdateRoom.mockResolvedValue({ data: null });

      // Mock console.error to avoid noise in test output
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

      renderWithMocks();

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith('Update failed: No data returned');
      });

      consoleSpy.mockRestore();
    });
  });

  describe('Cancel Functionality', () => {
    it('calls onOpenChange with false when cancel button is clicked', async () => {
      const user = userEvent.setup();
      renderWithMocks();

      const cancelButton = screen.getByText('Cancel');
      await user.click(cancelButton);

      expect(defaultProps.onOpenChange).toHaveBeenCalledWith(false);
    });

    it('does not call updateRoom when cancel is clicked', async () => {
      const user = userEvent.setup();
      renderWithMocks();

      const cancelButton = screen.getByText('Cancel');
      await user.click(cancelButton);

      expect(mockUpdateRoom).not.toHaveBeenCalled();
    });
  });

  describe('Loading States', () => {
    it('shows loading state on save button when mutation is loading', () => {
      // Test that the component renders without errors when loading is true
      // The actual loading state is tested through the mutation hook behavior
      renderWithMocks();
      expect(screen.getByText('Save Changes')).toBeInTheDocument();
    });

    it('disables buttons when loading', () => {
      // Test that buttons are present and can be interacted with
      renderWithMocks();
      const saveButton = screen.getByText('Save Changes');
      const cancelButton = screen.getByText('Cancel');

      expect(saveButton).toBeInTheDocument();
      expect(cancelButton).toBeInTheDocument();
    });
  });

  describe('Input Validation', () => {
    it('handles undefined form data fields', async () => {
      const user = userEvent.setup();
      const roomWithUndefinedFields = {
        id: 'room-1',
        name: undefined,
        pricePerNight: undefined,
        typePerson: undefined,
      };

      renderWithMocks({ ...defaultProps, room: roomWithUndefinedFields });

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              bedNumber: undefined,
              roomInformation: [],
              status: 'available',
            },
          },
        });
      });
    });

    it('handles empty string values', async () => {
      const user = userEvent.setup();
      const roomWithEmptyStrings = {
        id: 'room-1',
        name: '',
        pricePerNight: '',
        typePerson: '',
      };

      renderWithMocks({ ...defaultProps, room: roomWithEmptyStrings });

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalledWith({
          variables: {
            updateRoomId: 'room-1',
            input: {
              bedNumber: undefined,
              name: '',
              pricePerNight: '',
              roomInformation: [],
              status: 'available',
              typePerson: '',
            },
          },
        });
      });
    });
  });

  describe('Error Handling', () => {
    it('handles GraphQL errors', async () => {
      const user = userEvent.setup();
      const graphQLError = new Error('GraphQL error');
      graphQLError.graphQLErrors = [{ message: 'GraphQL error' }];
      mockUpdateRoom.mockRejectedValue(graphQLError);

      // Mock console.error to avoid noise in test output
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

      renderWithMocks();

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith('Error updating room:', graphQLError);
        expect(consoleSpy).toHaveBeenCalledWith('GraphQL errors:', [{ message: 'GraphQL error' }]);
      });

      consoleSpy.mockRestore();
    });

    it('handles network errors', async () => {
      const user = userEvent.setup();
      const networkError = new Error('Network error');
      networkError.networkError = { message: 'Network error' };
      mockUpdateRoom.mockRejectedValue(networkError);

      // Mock console.error to avoid noise in test output
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

      renderWithMocks();

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith('Error updating room:', networkError);
        expect(consoleSpy).toHaveBeenCalledWith('Network errors:', { message: 'Network error' });
      });

      consoleSpy.mockRestore();
    });
  });

  describe('Component Props', () => {
    it('handles missing refetch function', async () => {
      const user = userEvent.setup();
      const propsWithoutRefetch = { ...defaultProps, refetch: undefined };

      renderWithMocks(propsWithoutRefetch);

      const saveButton = screen.getByText('Save Changes');
      await user.click(saveButton);

      await waitFor(() => {
        expect(mockUpdateRoom).toHaveBeenCalled();
      });
    });

    it('handles missing onOpenChange function', () => {
      const propsWithoutOnOpenChange = { ...defaultProps, onOpenChange: undefined };

      // Should not throw error
      expect(() => renderWithMocks(propsWithoutOnOpenChange)).not.toThrow();
    });
  });

  describe('Accessibility', () => {
    it('has proper test IDs for testing', () => {
      renderWithMocks();

      expect(screen.getByTestId('edit-room-modal')).toBeInTheDocument();
      expect(screen.getByTestId('edit-room-modal-title')).toBeInTheDocument();
      expect(screen.getByTestId('edit-room-modal-content')).toBeInTheDocument();
      expect(screen.getByTestId('edit-room-modal-cancel')).toBeInTheDocument();
      expect(screen.getByTestId('edit-room-modal-save')).toBeInTheDocument();
    });

    it('has proper dialog attributes', () => {
      renderWithMocks();

      const dialogContent = screen.getByTestId('edit-room-modal');
      expect(dialogContent).toHaveClass('max-w-4xl', 'max-h-[90vh]', 'overflow-y-auto');
    });
  });

  describe('Snapshot Tests', () => {
    it('matches snapshot with basic section', () => {
      const { container } = renderWithMocks();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with amenities section', () => {
      const { container } = renderWithMocks({ ...defaultProps, section: 'amenities' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with images section', () => {
      const { container } = renderWithMocks({ ...defaultProps, section: 'images' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with details section', () => {
      const { container } = renderWithMocks({ ...defaultProps, section: 'details' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot when closed', () => {
      const { container } = renderWithMocks({ ...defaultProps, isOpen: false });
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
