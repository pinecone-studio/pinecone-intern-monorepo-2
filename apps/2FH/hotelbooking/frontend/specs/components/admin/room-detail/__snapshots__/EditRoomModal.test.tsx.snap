// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRoomModal Snapshot Tests matches snapshot when closed 1`] = `
<div
  data-open="false"
  data-testid="dialog"
>
  <div
    class="max-w-4xl max-h-[90vh] overflow-y-auto"
    data-testid="edit-room-modal"
  >
    <div
      data-testid="dialog-header"
    >
      <div
        data-testid="edit-room-modal-title"
      >
        Edit 
        Basic Information
      </div>
    </div>
    <div
      class="space-y-6 py-4"
    >
      <div
        class="min-h-[400px]"
        data-testid="edit-room-modal-content"
      >
        <div
          data-testid="basic-info-section"
        >
          <input
            data-testid="name-input"
            placeholder="Room name"
            value="Deluxe Suite"
          />
          <input
            data-testid="price-input"
            placeholder="Price per night"
            value="150000"
          />
          <input
            data-testid="type-person-input"
            placeholder="Type person"
            value="4"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-3 pt-4 border-t"
      data-testid="dialog-footer"
    >
      <button
        data-testid="edit-room-modal-cancel"
        variant="outline"
      >
        Cancel
      </button>
      <button
        class="bg-blue-600 hover:bg-blue-700"
        data-testid="edit-room-modal-save"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;

exports[`EditRoomModal Snapshot Tests matches snapshot with amenities section 1`] = `
<div
  data-open="true"
  data-testid="dialog"
>
  <div
    class="max-w-4xl max-h-[90vh] overflow-y-auto"
    data-testid="edit-room-modal"
  >
    <div
      data-testid="dialog-header"
    >
      <div
        data-testid="edit-room-modal-title"
      >
        Edit 
        Amenities
      </div>
    </div>
    <div
      class="space-y-6 py-4"
    >
      <div
        class="min-h-[400px]"
        data-testid="edit-room-modal-content"
      >
        <div
          data-testid="amenities-section"
        >
          <input
            data-testid="internet-input"
            placeholder="Internet"
            value="Free WiFi"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-3 pt-4 border-t"
      data-testid="dialog-footer"
    >
      <button
        data-testid="edit-room-modal-cancel"
        variant="outline"
      >
        Cancel
      </button>
      <button
        class="bg-blue-600 hover:bg-blue-700"
        data-testid="edit-room-modal-save"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;

exports[`EditRoomModal Snapshot Tests matches snapshot with basic section 1`] = `
<div
  data-open="true"
  data-testid="dialog"
>
  <div
    class="max-w-4xl max-h-[90vh] overflow-y-auto"
    data-testid="edit-room-modal"
  >
    <div
      data-testid="dialog-header"
    >
      <div
        data-testid="edit-room-modal-title"
      >
        Edit 
        Basic Information
      </div>
    </div>
    <div
      class="space-y-6 py-4"
    >
      <div
        class="min-h-[400px]"
        data-testid="edit-room-modal-content"
      >
        <div
          data-testid="basic-info-section"
        >
          <input
            data-testid="name-input"
            placeholder="Room name"
            value="Deluxe Suite"
          />
          <input
            data-testid="price-input"
            placeholder="Price per night"
            value="150000"
          />
          <input
            data-testid="type-person-input"
            placeholder="Type person"
            value="4"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-3 pt-4 border-t"
      data-testid="dialog-footer"
    >
      <button
        data-testid="edit-room-modal-cancel"
        variant="outline"
      >
        Cancel
      </button>
      <button
        class="bg-blue-600 hover:bg-blue-700"
        data-testid="edit-room-modal-save"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;

exports[`EditRoomModal Snapshot Tests matches snapshot with details section 1`] = `
<div
  data-open="true"
  data-testid="dialog"
>
  <div
    class="max-w-4xl max-h-[90vh] overflow-y-auto"
    data-testid="edit-room-modal"
  >
    <div
      data-testid="dialog-header"
    >
      <div
        data-testid="edit-room-modal-title"
      >
        Edit 
        Details
      </div>
    </div>
    <div
      class="space-y-6 py-4"
    >
      <div
        class="min-h-[400px]"
        data-testid="edit-room-modal-content"
      >
        <div
          data-testid="details-section"
        >
          <textarea
            data-testid="room-information-input"
            placeholder="Room information"
          >
            Free WiFi,Air Conditioning
          </textarea>
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-3 pt-4 border-t"
      data-testid="dialog-footer"
    >
      <button
        data-testid="edit-room-modal-cancel"
        variant="outline"
      >
        Cancel
      </button>
      <button
        class="bg-blue-600 hover:bg-blue-700"
        data-testid="edit-room-modal-save"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;

exports[`EditRoomModal Snapshot Tests matches snapshot with images section 1`] = `
<div
  data-open="true"
  data-testid="dialog"
>
  <div
    class="max-w-4xl max-h-[90vh] overflow-y-auto"
    data-testid="edit-room-modal"
  >
    <div
      data-testid="dialog-header"
    >
      <div
        data-testid="edit-room-modal-title"
      >
        Edit 
        Images
      </div>
    </div>
    <div
      class="space-y-6 py-4"
    >
      <div
        class="min-h-[400px]"
        data-testid="edit-room-modal-content"
      >
        <div
          data-testid="images-section"
        >
          <input
            data-testid="image-url-input"
            placeholder="Image URL"
            value="https://example.com/image1.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-3 pt-4 border-t"
      data-testid="dialog-footer"
    >
      <button
        data-testid="edit-room-modal-cancel"
        variant="outline"
      >
        Cancel
      </button>
      <button
        class="bg-blue-600 hover:bg-blue-700"
        data-testid="edit-room-modal-save"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;
